src/build/__init__.py

```diff
@@ -55,43 +55,47 @@ class IncompleteCheckWarning(Warning):
     """
 
 
-def check_version(requirement_string, extra=''):  # type: (str, str) -> bool
+def check_dependency(req_string, parent_extras=frozenset()):  # type: (str, AbstractSet[str]) -> Iterator[str]
     """
-    :param requirement_string: Requirement string
-    :param extra: Extra (eg. test in myproject[test])
+    Verify that a dependency and all of its dependencies are met.
+
+    :param req_string: Requirement string
+    :param parent_extras: Extras (eg. "test" in myproject[test])
+    :yields: Unmet dependencies
     """
-    import packaging.requirements
+    from packaging.requirements import Requirement
 
     if sys.version_info >= (3, 8):
         from importlib import metadata as importlib_metadata
     else:
         import importlib_metadata
 
-    req = packaging.requirements.Requirement(requirement_string)
-    env = {'extra': extra}
+    req = Requirement(req_string)
 
-    if req.marker and not req.marker.evaluate(env):
-        return True
+    if req.marker:
```

Comment:
Could we perhaps use somehow ``pip check`` instead?

Reply:
Well, that would place a runtime dependency on pip (we could use `ensurepip` I suppose but distros might then be inclined to patch `build` to 'unvendor' pip), and unless we want to be calling its internal API we're looking at a ~400ms overhead per check purely in import time.  IUC pip does not keep track of extras so `pip check` would not error for e.g. `setuptools_scm[toml]` if the toml extra dependencies weren't available.

Reply:
We already have pip as a runtime dependency - https://github.com/pypa/build/blob/master/src/build/env.py#L173-L188; and this would allow us to drop packaging as install dependency.

Reply:
Even then, we'd have to combine `check` with `list` or `show` to verify that the build dependencies are installed and `list` is the only one of those commands which produces machine-readable data.  I don't know how we'd verify that extras are satisfied with pip.

Reply:
The current runtime dependency on pip is for the environments, and it's not even required if ensurepip works. This adds a dependency on pip for `--no-isolation` which we most definitely do not want, that was the whole point of this routine.

Reply:
Seemed a worthwhile topic to bring up. Considering pip already has a checking functionality, perhaps we don't want to reinvent the wheel.